Better documentation for condition variables.

fpottier opened the pull request:
The specification has been made more precise:
- exactly one mutex per condition variable;
- [signal] and [broadcast] can be called only when the mutex is locked.
Furthermore, a longer high-level explanation of condition variables
has been added.

fpottier committed:
Better documentation for condition variables.

The specification has been made more precise:
- exactly one mutex per condition variable;
- [signal] and [broadcast] can be called only when the mutex is locked.
Furthermore, a longer high-level explanation of condition variables
has been added.

fpottier commented:
Ping @maranget : could you check whether this documentation seems reasonable to you? Thanks!

maranget commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
+   only when the property [P] is definitely true.
+   When a signal is received, however, one cannot assume that [P] is
+   definitely true; one must explicitly test whether [P] is true.
+   The reason is that,
+   between the moment when the signal is sent
+   and the moment when a waiting thread receives the signal
+   and is scheduled,
+   the property [P] may be falsified by some other thread
+   that is able to acquire the mutex [m] and alter the data structure [D].
```
Program logic can, in some situations (single writer, single reader,  for instance) guarantee that [P} is true when the waiting thread wakes up. However, there is another reason for checking [P]: a waiting thread can wake up for no reason (_spurious wakeup_).

fpottier replied:
OK, I have added a mention of spurious wakeups.

maranget commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
+     if (* the property P over D is now satisfied *) then Condition.signal c;
      Mutex.unlock m
    ]}
+
+   {!signal} and {!broadcast} should be called only when the mutex [m]
+   associated with the condition variable [c] is locked.
+   Otherwise, there would be a risk that some other thread is currently
+   inside a critical section, has determined that the property [P]
+   associated with [c] is false, but has not yet called [wait]. In such a
+   case, the signal would be lost, leading to a potential deadlock.
```
I am not so sure about this, some code sometimes calls {!signal} and {!broadcast} from outside of critical sections. Some ["officicial"  description](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_signal.html) of POSIX signal fonction  contradicts the claim above:

> The pthread_cond_broadcast() or pthread_cond_signal() functions may be called by a thread whether or not it currently owns the mutex that threads calling pthread_cond_wait() or pthread_cond_timedwait() have associated with the condition variable during their waits; however, if predictable scheduling behavior is required, then that mutex shall be locked by the thread calling pthread_cond_broadcast() or pthread_cond_signal().

As far as I understand it, the implementation can take care of the involved mutex, as it is passed as an argument to wait. Also notice that, in POSIX threads, it is an error to invoke {! wait} on the same condition variable with different mutexes. See the error section  [here](https://linux.die.net/man/3/pthread_cond_wait).

However, it is recommended practice to set [P] and signal "atomically", _i.e._ in critical section. This leads to clearer code. It may also be that some coding is more efficient than the other, depending upon scheduler. I am not sure that such a discussion is adequate here...

xavierleroy replied:
POSIX threads definitely support using signal and broadcast outside of the critical section, and it's not clear to me how this can lead to a deadlock.  However, common wisdom is that signaling inside the critical section leads to better scheduling and more fair executions overall.

fpottier replied:
If a thread A uses `signal` or `broadcast` without holding the lock, then one runs the risk that a thread B might be inside a critical section, immediately after determining that the desired property is false, and before calling `wait`. Because no thread is currently waiting, the signal is lost. Then, thread B goes to sleep, even though the desired condition is in fact true. This can lead to a situation such as a consumer waiting indefinitely to extract an item out of a queue, when the queue is in fact nonempty: this is what I call a deadlock.

Furthermore, it seems dubious to call `signal` or `broadcast` outside a critical section, since one cannot know in that case whether the desired property is true.

Finally, I note that in Java, `notify` and `notifyAll` can be called only when the lock is held.

This is why I suggest requiring that the lock be held when `signal` or `broadcast` is called. But, if you insist, I can change this to a recommendation instead of a requirement.

maranget replied:
> If a thread A uses `signal` or `broadcast` without holding the lock, then one runs the risk that a thread B might be inside a critical section, immediately after determining that the desired property is false, and before calling `wait`. Because no thread is currently waiting, the signal is lost. Then, thread B goes to sleep, even though the desired condition is in fact true. This can lead to a situation such as a consumer waiting indefinitely to extract an item out of a queue, when the queue is in fact nonempty: this is what I call a deadlock.

You have to describe a complete scenario to convince us . For instance, I do not understand why loosing signal here  is  a problem. The fact is that POSIX programmers agree that signaling from outside of critical section is possible, although not recommended. And again, one can signal when the property is false, even inside critical sections (see comment below).

fpottier replied:
Actually, my scenario is complete already: thread A signals, the signal is lost, thread B waits indefinitely for a signal, even though the property desired by thread B is already true.

I agree that signalling outside a critical section does not compromise *safety*, but from the point of view of *liveness*, a signal sent outside of a critical section can be lost and therefore is just as good as no signal at all. In other words, when proving absence of deadlocks, a call to `signal` or `broadcast` outside of a critical section cannot contribute to the proof.

This is why I suggest to at least *strongly recommend* to use these functions inside a critical section.

maranget replied:
May I beg to differ: you assume some implementation of {! signal} that is not described. I am not sure that a POSIX compliant implementation could legitimately lose the signal here, although such implementation does not look that easy to design at first sight.

fpottier replied:
Actually, this version of the [standard](https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_cond_signal.html) says: *the `pthread_cond_signal()` and `pthread_cond_broadcast()` functions have no effect if there are no threads currently blocked on cond.* This suggests that the signal is lost if no thread is currently waiting. (And it seems to be common wisdom that this problem exists; it is known as a *lost wakeup*.)

The same document also indicates that it is OK to use `signal` and `broadcast` without holding the mutex, yet it adds: *however, if predictable scheduling behaviour is required, then the mutex is locked by the thread calling `signal` or `broadcast`*. It is unclear what this means...

maranget replied:
Thanks for the "lost wakeup", I did not know that this situation has a name. Looking for it on the web I [found](https://docs.oracle.com/cd/E19120-01/open.solaris/816-5137/sync-30/index.html):

> This can occur only if the condition being tested is modified without holding the mutex lock associated with the condition. As long as the condition being tested is modified only while holding the associated mutex, pthread_cond_signal() and pthread_cond_broadcast() can be called regardless of whether they are holding the associated mutex.

For what it's worth, I would interpret that lost wakeup may occur in racy programs. As far as POSIX is concerned, anything can happen.  For OCaml, the situation may be different, as racy programs are allowed.

My current opinion  is that the question is a bit too complex to be answered in `condition.mli` comments, and that I'll continue to call {! signal} inside critical sections. I would tend to recommend avoiding the practice of signalling from outside critical sections, more for code clarity than anything else.

fpottier replied:
I have thought about the sentence that you quote (*As long as the condition being tested is modified only while holding the associated mutex, [...]*). I believe it may be true, although I am unable to construct a clear argument. (I think it is time I read [this paper](https://people.cs.kuleuven.be/~bart.jacobs/esop18.pdf)...) Anyway, I will change the text to remove the requirement that the lock be held.

maranget commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
-   condition variable [c]. *)
+(**[signal c] wakes up one of the threads waiting on the condition
+   variable [c]. This call is permitted only if the mutex [m]
+   associated with [c] is currently locked. It should be the case that
+   the property [P] associated with the condition variable [c] is true
+   when [signal] is called. *)
```
Sentence "It should be ... is called" is not clear. In fact as it has always been mandated to  check  [P] after having returned from suspension, there is nothing dangerous in signaling without checking [P]. 

Moreover when implementing a concurrent FIFO, there are two correct alternatives for signaling the threads that are waiting for the queue to become non-empty:
1. 'put' calls {! broadcast} when it puts some element in an empty queue.
2. 'put' calls {! signal} in all situations.

Herlihy and Shavit' "The art of multiprocessor programming" uses technique 2 in its FIFO implementation (Sec. 8.2.2), thereby contradicting the comment. Edit "-> contradicting a naive reading of the comment"

fpottier replied:
I have removed the indication that `P` should be true when `signal` or `broadcast` is called.

xavierleroy commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
      done;
      (* Modify D *)
-     if (* the predicate P over D is now satisfied *) then Condition.signal c;
+     if (* the property P over D is now satisfied *) then Condition.signal c;
```
I'm not sure this is a good example...  In typical examples like bounded buffers, waiting and signaling are done on different conditions, and signaling is not guarded by an "if":
```
  while not (is_empty q) do Condition.wait q.nonempty m done;
  let x = take q in
  Condition.signal q.notfull;
  x
```
I can't think of a good example that follows the "one condition" pattern given in the docs.

Why not show the bounded buffer example, while we're at it?

fpottier replied:
Indeed, the code snippet (which I did not change) is weird, as it tries to illustrate both sides (waiting and signaling) using a single code fragment and a single condition variable. I will change it to show two separate snippets taken from the bounded buffer example.

maranget commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
-(** [signal c] restarts one of the processes waiting on the
-   condition variable [c]. *)
+(**[signal c] wakes up one of the threads waiting on the condition
+   variable [c]. This call is permitted only if the mutex [m]
```
Again , one may call {! signal} without having locked the mutex.

fpottier replied:
I have removed this requirement.

xavierleroy commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
-(** [signal c] restarts one of the processes waiting on the
-   condition variable [c]. *)
+(**[signal c] wakes up one of the threads waiting on the condition
+   variable [c]. This call is permitted only if the mutex [m]
+   associated with [c] is currently locked. It should be the case that
+   the property [P] associated with the condition variable [c] is true
+   when [signal] is called. *)
```
I don't think this is right.  Since the waiting threads will all re-test the property P when woken up, the condition variable can be signaled at any time and in any state of the shared data.  Moreover, between the time thread 1 signals the condition and unlocks the mutex, and the time thread 2 is woken up from the wait and reacquires the mutex, a thread 3 may have acquired the mutex, falsified property P, and released the mutex.

So, signaling "only when the property P holds" is an optimization, to avoid too many useless wake-ups of waiting threads, but not a strong requirement.

fpottier replied:
I have removed this requirement.

xavierleroy reviewed (commented):
Thanks for the additional documentation.  Some comments on signal and broadcast below.

xavierleroy commented:
I see @maranget and I are in violent agreement here !

maranget commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
+   with the condition variable [c], and only if [m] is currently locked.
+   This call atomically unlocks the mutex [m] and suspends the
+   current thread on the condition variable [c]. This thread can
+   later be woken up after the condition variable [c] has been signaled
+   via {!signal} or {!broadcast}; however, it can also be woken up for
+   no reason. The mutex [m] is locked again before [wait] returns. One
+   cannot assume that the property [P] associated with the condition
+   variable [c] holds when [wait] returns; one must explicitly test
+   whether [P] holds after calling [wait]. *)
```
The second part of the paragraph, starting from "This thread can be woken.." is redundant w.r.t. the introduction and with {! signal} description.

fpottier replied:
A little redundancy does not hurt, I think.

xavierleroy commented:
More thoughts.  Maybe the simplest way to explain condition variables is to say they provide an efficient alternative to busy waiting: rather than writing
```
  lock m; while not P do (unlock m; lock m) done; do_something; unlock m
```
one writes
```
  lock m; while not P do wait c m done; do_something; unlock m
```
and makes sure that c is signaled whenever something changes that could make P true.

maranget commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,112 @@
-   condition variable [c]. *)
+(**[broadcast c] wakes up all threads waiting on the condition
+   variable [c]. This call is permitted only if the mutex [m]
+   associated with [c] is currently locked. It should be the case that
+   the property [P] associated with the condition variable [c] is true
+   when [broadcast] is called. *)
```
See comment on signal.

maranget reviewed (approved):
I have commented comments.

maranget commented:
A simple example (code not tested):
``` OCaml
type 'a buff =
  {  queue : 'a Queue.t ;
     mtx : Mutex.t ;
     isEmpty : Condition.t ;
  }

let create () =
  { queue=Queue.create () ; mtx=Mutex.create () ;
    isEmpty=Condition.create ();
  }


let add v q =
  Mutex.lock q.mtx ;
  let wasEmpty = Queue.is_empty q.queue in
  Queue.add v q.queue ;
  if wasEmpty then Condition.broadcast q.isEmpty ;
  Mutex.unlock q.mtx

let take q =
  Mutex.lock q.mtx ;
  while Queue.is_empty q.queue do Condition.wait q.isEmpty q.mtx done ;
  let r = try Queue.take q.queue with Queue.Empty -> assert false in
  Mutex.unlock q.mtx ;
  r
```

fpottier committed:
Mention spurious wakeups.

fpottier committed:
Instead of requiring that [signal] and [broadcast] be called inside
a critical section, make this a recommendation. Also, remove the
(informal) requirement that the property [P] should be true when
a signal is sent.

fpottier committed:
Present condition variables as an efficient alternative to busy-waiting.

fpottier committed:
Add a complete example of an unbounded queue.

fpottier commented:
I have added Luc's example of the unbounded queue (with some renamings). It is interesting to note that in this example, because `broadcast` is called inside the critical section, it is (relatively) easy to see that a simple invariant holds: *if the queue is nonempty, then no thread is waiting*.

It would also be interesting to note that replacing `broadcast` with `signal` would clearly break this invariant (and would in fact make the code incorrect), but I have not added this comment.

Using `broadcast` outside of the critical section would perhaps be correct (I am willing to believe it) but would destroy this simple invariant, making the correctness argument more complex. I have not added this comment either.

@maranget @xavierleroy I think the file is ready for you to have another look, if you would like. Thanks!

fpottier committed:
Avoid a long line.

xavierleroy commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,164 @@
+   Calling {!wait} allows the waiting thread to be suspended, so it
+   does not consume any computing resources while waiting.
+
+   With a condition variable [c], exactly one mutex [m] is associated.
```
Yoda speak this is a bit of, no?
"Exactly one mutex  [m] is associated with a (given) condition variable [c]."
(I see there are more occurrences of the pattern below, so if you like the pattern it's fine with me too.)

fpottier replied:
I didn't pay attention, but it sounds natural to me...

wikku replied:
"A condition variable has exactly one associated mutex" ?

fpottier replied:
I feel that my formulation is the most precise/unambiguous I can think of.

xavierleroy reviewed (approved):
Very nice documentation, thanks a lot !  Tiny style remark below, can be ignored.

wikku commented on stdlib/condition.mli:
```diff
@@ -13,41 +13,164 @@
+   condition variable is used for this purpose.
+
+   In short, a condition variable [c] is used to convey the information
+   that a certain property [P] about a shared data structure [D],
```
I think italics for P and D would be more fitting, they exist in the programmer's mind and don't have to be reflected in source code.

fpottier replied:
OK, changed to italics.

fpottier committed:
Use italics for P and D.

fpottier committed:
Add a Changes entry.

fpottier committed:
Move the Changes entry.

xavierleroy commented:
Let's merge, there's no reason to wait longer.
